### PR TITLE
Fix upgrade module to v4

### PR DESCRIPTION
**Pull request by jf-viguier**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Add required hooks and configuration values for v4 upgrade
| Type?         | bug fix
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/23218
| How to test?  | see issue

It seems that the upgrade file is not launched at prestashop 1.7 upgrade in some cases. So I added a default commentsPerPage in paginate method

### CODE REVIEW

File: src/Repository/ProductCommentRepository.php [hunk 1]
```diff
@@ -79,6 +79,9 @@ public function __construct(
      */
     public function paginate($productId, $page, $commentsPerPage, $validatedOnly)
     {
+        if (empty($commentsPerPage)) {
+            $commentsPerPage = 5;
```
Comment (Progi1984) [inline src/Repository/ProductCommentRepository.php:83]: Do you put it in a constant, please ?

Comment (jf-viguier) [reply src/Repository/ProductCommentRepository.php]: done

**Review by Progi1984 — APPROVED** [review]

**Comment by jf-viguier** [comment]
thanks @Progi1984, can you add my 3 pr in this module to 4.2.2 version ?
This pr should fix this pr too https://github.com/PrestaShop/productcomments/pull/92
(in fact it's just a test to make)

File: upgrade/install-4.0.0.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $res &= (bool) Configuration::set('PRODUCT_COMMENTS_COMMENTS_PER_PAGE', 5);
+    }
+    if (!Configuration::hasKey('PRODUCT_COMMENTS_USEFULNESS')) {
+        $res &= (bool) Configuration::set('PRODUCT_COMMENTS_USEFULNESS', 1);
+    }
+    $res &= (bool) $object->unregisterHook('displayRightColumnProduct');
+    $res &= (bool) $object->registerHook('displayProductAdditionalInfo');
+    $res &= (bool) $object->registerHook('displayFooterProduct');
+
+    return $res;
```
Comment (PierreRambaud) [inline upgrade/install-4.0.0.php:43]:
```suggestion
    return (bool) $res;
```

File: upgrade/install-4.0.0.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+function upgrade_module_4_0_0($object)
+{
+    $res = true;
+    if (!Configuration::hasKey('PRODUCT_COMMENTS_COMMENTS_PER_PAGE')) {
+        $res &= (bool) Configuration::set('PRODUCT_COMMENTS_COMMENTS_PER_PAGE', 5);
```
Comment (kpodemski) [inline upgrade/install-4.0.0.php:34]: @jf-viguier using `set` doesn't make any sense, `set` in legacy code is used in some specific cases, not for a real update of the value, you should go for `Configuration::updateValue`

```suggestion
        $res &= (bool) Configuration::updateValue('PRODUCT_COMMENTS_COMMENTS_PER_PAGE', 5);
```

File: upgrade/install-4.0.0.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    if (!Configuration::hasKey('PRODUCT_COMMENTS_USEFULNESS')) {
+        $res &= (bool) Configuration::set('PRODUCT_COMMENTS_USEFULNESS', 1);
```
Comment (kpodemski) [inline upgrade/install-4.0.0.php:37]:
```suggestion
        $res &= (bool) Configuration::updateValue('PRODUCT_COMMENTS_USEFULNESS', 1);
```

**Review by kpodemski — CHANGES_REQUESTED** [review]

File: upgrade/install-4.0.0.php (hunk 3 above)
Comment (jf-viguier) [reply upgrade/install-4.0.0.php]: Thanks @kpodemski, fixed

File: upgrade/install-4.0.0.php (hunk 4 above)
Comment (jf-viguier) [reply upgrade/install-4.0.0.php]: fixed too

**Review by matthieu-rolland — APPROVED** [review]
@kpodemski seems like your feedback has been dealt with 👍

**Comment by jf-viguier** [comment]
@kpodemski is it ok now ?

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
@jf-viguier great, thanks 👍

**Review by atomiix — APPROVED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @jf-viguier 

I'm really sorry but it seems I have an issue with installing the productcomment module v3.6.1 on 1.6.1.20

image

I tried many times but in vain :confounded: 
Ping @PrestaShop/qa-functional could someone please test this PR cause I can't.

Thanks in advance :pray:

**Comment by khouloudbelguith** [comment]
Hi @jf-viguier,

I followed these steps:

1. Install PrestaShop 1.6.1.24
2. Install module product comment 3.6.0 from GitHub
3. Add a comment with guest and customer
4. Approve those comments
5. Upgrade PrestaShop to PS1.7.7.5 (I found that the productcomment module is upgraded to v4.2.1)
6. I Upgraded manually the product comment module to this PR (v4.2.2)
7. Go to FO
image

8. View a product with a comment
9. See error => No comment is displayed 
image

Thanks to check and feedback.

**Comment by jf-viguier** [comment]
@khouloudbelguith have you cleared the ps cache ?

**Comment by sarahdib** [comment]
@jf-viguier I have the same bug as @khouloudbelguith  and all cache are cleared.

**Comment by matks** [comment]
> There is an extra bug I guess

I'm note sure to understand 😅  so is it in the scope of the contribution ?

**Comment by jf-viguier** [comment]
> so is it in the scope of the contribution ?

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/prestashop-core-developers is there any one have an idea about how to test this PR :pray: 
I suggest these steps, so can someone check if these are correcte:

1. Install PS 1.6.1.24
2. Install Product comments with these cmds:

- `cd /your-shop/modules/`
- `git clone https://github.com/PrestaShop/productcomments.git`
- `git checkout tags/3.6.0 -b 3.6.0`
- **BO**>**Module Catalog**>Click on **install** of **Product Comments**
- `git fetch origin pull/107/head:pr-107 && git checkout pr-107 && git reset --hard && git clean -fd && git rebase origin/dev && composer install`

3. Create 6 comments with guests and Customers
4. Approve all
5. Upgrade the shop to 1.7.8.2
6. Check the comments in FO

Thanks in advance :pray:

**Comment by jf-viguier** [comment]
The productcomments is updated during pretashop upgrade from addons : so the code of this pr will not be launched, how to solve this @kpodemski ?

**Comment by kpodemski** [comment]
@jf-viguier hmm, you're right

I think we need to create a package for the module.
1. `composer install`
2. Make a .zip from the `productcomments` directory. And install this module after the upgrade.

@jf-viguier that should do it, right?

**Comment by jf-viguier** [comment]
@kpodemski I guess no, upgrade/install-4.0.0.php of my pr has to be launched. Prestashop upgrade will upgrade the module to the version 5.

**Comment by Progi1984** [comment]
Discussed with @hibatallahAouadni about the process, she will test with this process. Thanks to her.

**Review by hibatallahAouadni — CHANGES_REQUESTED** [review]
Hello everyone,

Following my steps (mentioned above), I encounter first issue while upgrading:

image

Then, I give the access rights to the module and re-upgrade to 1.7.8.2, and it was a success:

image

As you can see in module manager, the PR is well installed:

image

Clicking on Configure, you could see the 6 created comments before upgrade:

image

But on FO, **the block product comments is not displayed**:

image

I checked the position of hooks and it seems good to me:

image

I check the table of product_comments, and it seems all good to me:

image

So, @jf-viguier you need to follow my steps to check and fix the issue (the block product comments is not displayed).
Please check and feedback.

Thanks!

**Comment by jf-viguier** [comment]
modules\productcomments\upgrade\install-4.0.0.php is well executed during test steps.

There are two issues with your steps @hibatallahAouadni 
1) After prestashop upgrade, there are some missing hooks for the module (displayRightColumnProduct & displayFooterProduct):
image

2) my changes in modules\productcomments\src\Repository\ProductCommentRepository.php are deleted by prestashop upgrade, so you have to update it manually

There is still a javascript error to fix in modules\productcomments\views\templates\hook\product-additional-info-quickview.tpl (remove all script here) and then comments will whow up.

Please check @hibatallahAouadni

**Comment by jf-viguier** [comment]
I've just rebased my pr

**Comment by hibatallahAouadni** [comment]
Hello @jf-viguier 

After adding this hook `displayFooterProduct`:

image

All comments are well displayed and the pagination works :100: 

https://user-images.githubusercontent.com/23308427/149790031-3dcc800a-19a0-42e6-9eef-ec2df259f238.mp4

LGTM, QA :heavy_check_mark: 

Thanks for your hard working :rocket:

**Review by hibatallahAouadni — APPROVED** [review]
Hello @jf-viguier 

After adding this hook `displayFooterProduct`:

image

All comments are well displayed and the pagination works :100: 

https://user-images.githubusercontent.com/23308427/149790031-3dcc800a-19a0-42e6-9eef-ec2df259f238.mp4

LGTM, QA :heavy_check_mark: 

Thanks for your hard working :rocket:

**Comment by jf-viguier** [comment]
Thanks @hibatallahAouadni

**Comment by Progi1984** [comment]
Thanks @jf-viguier & @hibatallahAouadni (for the hard work QA on this PR).

**Pull request merged**